File: src/Web/TypeScriptTS/proxy.ts
interface DocumentStore {
  get(id: number): string;
}

class RemoteDocumentStore implements DocumentStore {
  fetchCount = 0;

  get(id: number): string {
    this.fetchCount += 1;
    return `doc(${id})`;
  }
}

class DocumentStoreProxy implements DocumentStore {
  private backend: RemoteDocumentStore | undefined;
  private readonly cache = new Map<number, string>();

  get backendCount(): number {
    return this.backend === undefined ? 0 : 1;
  }

  get fetchCount(): number {
    return this.backend === undefined ? 0 : this.backend.fetchCount;
  }

  get(id: number): string {
    const cached = this.cache.get(id);
    if (cached !== undefined) return cached;
    if (this.backend === undefined) this.backend = new RemoteDocumentStore();
    const value = this.backend.get(id);
    this.cache.set(id, value);
    return value;
  }
}

const store = new DocumentStoreProxy();
const first = store.get(42);
const second = store.get(42);
console.log(`backend=${store.backendCount};fetches=${store.fetchCount};first=${first};second=${second}`);

File: src/Other/MicroPython/proxy.py
class RemoteDocumentStore:
    def __init__(self):
        self.fetches = 0

    def get(self, document_id):
        self.fetches += 1
        return "doc({})".format(document_id)


class DocumentStoreProxy:
    def __init__(self):
        self.backend = None
        self.backend_creations = 0
        self.cache = {}

    def get(self, document_id):
        if document_id in self.cache:
            return self.cache[document_id]
        if self.backend is None:
            self.backend = RemoteDocumentStore()
            self.backend_creations += 1
        value = self.backend.get(document_id)
        self.cache[document_id] = value
        return value


store = DocumentStoreProxy()
first = store.get(42)
second = store.get(42)
assert store.backend is not None
print("backend={};fetches={};first={};second={}".format(
    store.backend_creations, store.backend.fetches, first, second))
